File: src/main/java/me/crylonz/DeadChestListener.java
package me.crylonz;

import org.bukkit.*;
import org.bukkit.block.Block;
import org.bukkit.block.BlockFace;
import org.bukkit.block.BlockState;
import org.bukkit.block.Skull;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.entity.ArmorStand;
import org.bukkit.entity.EntityType;
import org.bukkit.entity.Player;
import org.bukkit.event.Event;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.block.*;
import org.bukkit.event.entity.EntityExplodeEvent;
import org.bukkit.event.entity.PlayerDeathEvent;
import org.bukkit.event.player.PlayerArmorStandManipulateEvent;
import org.bukkit.event.player.PlayerInteractEvent;
import org.bukkit.inventory.ItemStack;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import static me.crylonz.DeadChest.*;
import static me.crylonz.DeadChestManager.generateHologram;
import static me.crylonz.DeadChestManager.playerDeadChestAmount;
import static me.crylonz.Utils.*;

public class DeadChestListener implements Listener {

    @EventHandler(priority = EventPriority.HIGH)
    public void onPlayerDeathEvent(PlayerDeathEvent e) {

        if (e.getKeepInventory()) {
            return;
        }

        Player p = e.getEntity().getPlayer();

        if (p == null
                || excludedWorlds.contains(p.getWorld().getName())
                || (!generateDeadChestInCreative) && p.getGameMode().equals(GameMode.CREATIVE)) {
            return;
        }

        if (worldGuardCheck(p) && (p.hasPermission("deadchest.generate") || !requirePermissionToGenerate)) {
            if ((playerDeadChestAmount(p) < maxDeadChestPerPlayer ||
                    maxDeadChestPerPlayer == 0) && p.getMetadata("NPC").isEmpty()) {

                World world = p.getWorld();
                Location loc = p.getLocation();

                if (!generateOnLava && loc.getBlock().getType().equals(Material.LAVA)) {
                    generateLog("Player dies in lava : No deadchest generated");
                    return;
                }

                if (!generateOnWater && loc.getBlock().getType().equals(Material.WATER)) {
                    generateLog("Player dies in water : No deadchest generated");
                    return;
                }

                if (!generateOnRails &&
                        loc.getBlock().getType().equals(Material.RAIL) ||
                        loc.getBlock().getType().equals(Material.ACTIVATOR_RAIL) ||
                        loc.getBlock().getType().equals(Material.DETECTOR_RAIL) ||
                        loc.getBlock().getType().equals(Material.POWERED_RAIL)) {
                    generateLog("Player dies on rails : No deadchest generated");
                    return;
                }

                if (!generateInMinecart && p.getVehicle() != null) {
                    if (p.getVehicle().getType().equals(EntityType.MINECART)) {
                        generateLog("Player dies in a minecart : No deadchest generated");
                        return;
                    }
                }


                // Handle case bottom of the world
                if (loc.getY() < 1) {
                    loc.setY(world.getHighestBlockYAt((int) loc.getX(), (int) loc.getZ()) + 1);
                    if (loc.getY() < 1)
                        loc.setY(1);
                }

                // Handle case top of the world
                else if (loc.getBlockY() >= world.getMaxHeight()) {

                    int y = world.getMaxHeight() - 1;
                    loc.setY(y);

                    while (world.getBlockAt(loc).getType() != Material.AIR && y > 0) {
                        y--;
                        loc.setY(y);
                    }

                    if (y < 1) {
                        p.sendMessage(local.get("loc_prefix") + local.get("loc_noDCG"));
                        return;
                    }
                }
                // Handle standard case
                else {

                    if (world.getBlockAt(loc).getType() == Material.DARK_OAK_DOOR ||
                            world.getBlockAt(loc).getType() == Material.ACACIA_DOOR ||
                            world.getBlockAt(loc).getType() == Material.BIRCH_DOOR ||
                            (!Utils.isBefore1_16() && world.getBlockAt(loc).getType() == Material.CRIMSON_DOOR) ||
                            world.getBlockAt(loc).getType() == Material.IRON_DOOR ||
                            world.getBlockAt(loc).getType() == Material.JUNGLE_DOOR ||
                            world.getBlockAt(loc).getType() == Material.OAK_DOOR ||
                            world.getBlockAt(loc).getType() == Material.SPRUCE_DOOR ||
                            (!Utils.isBefore1_16() && world.getBlockAt(loc).getType() == Material.WARPED_DOOR) ||
                            world.getBlockAt(loc).getType() == Material.VINE ||
                            world.getBlockAt(loc).getType() == Material.LADDER) {
                        Location tmpLoc = getFreeBlockAroundThisPlace(world, loc);

                        if (tmpLoc != null) {
                            loc = tmpLoc;
                        }
                    }

                    if (world.getBlockAt(loc).getType() != Material.AIR
                            && world.getBlockAt(loc).getType() != Material.CAVE_AIR
                            && world.getBlockAt(loc).getType() != Material.VOID_AIR
                            && world.getBlockAt(loc).getType() != Material.WATER
                            && world.getBlockAt(loc).getType() != Material.LAVA) {

                        while (world.getBlockAt(loc).getType() != Material.AIR &&
                                loc.getY() < world.getMaxHeight()) {
                            loc.setY(loc.getY() + 1);
                        }
                    }
                }

                Location groundLocation = loc.clone();
                groundLocation.setY(groundLocation.getY() - 1);
                if (isBefore1_17() && world.getBlockAt(groundLocation).getType() == Material.valueOf("GRASS_PATH") ||
                        !isBefore1_17() && world.getBlockAt(groundLocation).getType() == Material.DIRT_PATH ||
                        world.getBlockAt(groundLocation).getType() == Material.FARMLAND) {
                    loc.setY(loc.getY() + 1);
                }

                Block b = world.getBlockAt(loc);

                if (!isInventoryEmpty(p.getInventory())) {

                    switch (dropBlock) {
                        case 2:
                            b.setType(Material.PLAYER_HEAD);
                            BlockState state = b.getState();
                            Skull skull = (Skull) state;
                            skull.setOwningPlayer(p);
                            skull.update();
                            break;
                        case 3:
                            b.setType(Material.BARREL);
                            break;
                        case 4:
                            b.setType(Material.SHULKER_BOX);
                            break;
                        case 5:
                            b.setType(Material.ENDER_CHEST);
                            break;
                        default:
                            b.setType(Material.CHEST);
                    }

                    String firstLine = local.replacePlayer(local.get("holo_owner"), e.getEntity().getDisplayName());
                    ArmorStand holoName = generateHologram(b.getLocation(), firstLine, 0.5f, -0.95f, 0.5f, false);

                    String secondLine = local.get("holo_loading");
                    ArmorStand holoTime = generateHologram(b.getLocation(), secondLine, 0.5f, -1.2f, 0.5f, true);

                    // Remove items with curse of vanishing
                    for (ItemStack is : p.getInventory().getContents()) {
                        if (is != null && is.getEnchantments().containsKey(Enchantment.VANISHING_CURSE)) {
                            p.getInventory().remove(is);
                        }
                    }

                    if (p.getInventory().getHelmet() != null
                            && p.getInventory().getHelmet().getEnchantments().containsKey(Enchantment.VANISHING_CURSE)) {
                        p.getInventory().setHelmet(null);
                    }

                    if (p.getInventory().getChestplate() != null
                            && p.getInventory().getChestplate().getEnchantments().containsKey(Enchantment.VANISHING_CURSE)) {
                        p.getInventory().setChestplate(null);
                    }

                    if (p.getInventory().getLeggings() != null
                            && p.getInventory().getLeggings().getEnchantments().containsKey(Enchantment.VANISHING_CURSE)) {
                        p.getInventory().setLeggings(null);
                    }

                    if (p.getInventory().getBoots() != null
                            && p.getInventory().getBoots().getEnchantments().containsKey(Enchantment.VANISHING_CURSE)) {
                        p.getInventory().setBoots(null);
                    }

                    if (p.getInventory().getItemInOffHand().getEnchantments().containsKey(Enchantment.VANISHING_CURSE)) {
                        p.getInventory().setItemInOffHand(null);
                    }

                    for (String item : excludedItems) {
                        if (item != null && Material.getMaterial(item.toUpperCase()) != null) {
                            p.getInventory().remove(Material.getMaterial(item.toUpperCase()));
                        }
                    }

                    chestData.add(new ChestData(p.getInventory(), b.getLocation(), p, p.hasPermission("deadChest.infinityChest"), holoTime, holoName));

                    e.getDrops().clear();
                    p.getInventory().clear();

                    if (displayDeadChestPositionOnDeath) {
                        p.sendMessage(local.get("loc_prefix") + local.get("loc_chestPos") + " X: " +
                                ChatColor.WHITE + b.getX() + ChatColor.GOLD + " Y: " +
                                ChatColor.WHITE + b.getY() + ChatColor.GOLD + " Z: " +
                                ChatColor.WHITE + b.getZ());
                    }

                    fileManager.saveModification();
                    generateLog("New deadchest for [" + p.getName() + "] in " + b.getWorld().getName() + " at X:" + b.getX() + " Y:" + b.getY() + " Z:" + b.getZ());

                    if (logDeadChestOnConsole)
                        log.info("New deadchest for [" + p.getName() + "] at X:" + b.getX() + " Y:" + b.getY() + " Z:" + b.getZ());
                } else {
                    generateLog("Player [" + p.getName() + "] died without inventory : No Deadchest generated");
                }
            }
        }
    }

    @EventHandler
    public void onClick(PlayerInteractEvent e) {

        Block block = e.getClickedBlock();

        if (block != null && isGraveBlock(block.getType())) {
            // if block is a dead chest
            if (e.getAction() == Action.LEFT_CLICK_BLOCK || e.getAction() == Action.RIGHT_CLICK_BLOCK) {

                for (ChestData cd : chestData) {
                    if (cd.getChestLocation().equals(block.getLocation())) {
                        // if everybody can open chest or if the chest is the chest of the current player
                        if (!OnlyOwnerCanOpenDeadChest || e.getPlayer().getUniqueId().toString().equals(cd.getPlayerUUID())
                                || e.getPlayer().hasPermission("deadChest.chestPass")) {

                            if (!e.getPlayer().hasPermission("deadchest.get") && requirePermissionToGetChest) {
                                generateLog(String.format("Player [%s] need to have deadchest.get permission to generate", e.getPlayer().getName()));
                                e.getPlayer().sendMessage(local.get("loc_prefix") + local.get("loc_noPermsToGet"));
                                e.setCancelled(true);
                                return;
                            }

                            generateLog("Deadchest of [" + cd.getPlayerName() + "] was taken by [" + e.getPlayer().getName() + "] in " + e.getPlayer().getWorld().getName());

                            // put all item on the inventory
                            if (dropMode == 1) {
                                for (ItemStack i : cd.getInventory()) {
                                    if (i != null) {

                                        if ((i.getType() == Material.IRON_HELMET ||
                                                i.getType() == Material.GOLDEN_HELMET ||
                                                i.getType() == Material.LEATHER_HELMET ||
                                                i.getType() == Material.DIAMOND_HELMET ||
                                                i.getType() == Material.CHAINMAIL_HELMET ||
                                                i.getType() == Material.TURTLE_HELMET ||
                                                (!Utils.isBefore1_16() && i.getType() == Material.NETHERITE_HELMET)) &&
                                                !i.getEnchantments().containsKey(Enchantment.BINDING_CURSE) &&
                                                e.getPlayer().getInventory().getHelmet() == null)
                                            e.getPlayer().getInventory().setHelmet(i);

                                        else if ((i.getType() == Material.IRON_BOOTS ||
                                                i.getType() == Material.GOLDEN_BOOTS ||
                                                i.getType() == Material.LEATHER_BOOTS ||
                                                i.getType() == Material.DIAMOND_BOOTS ||
                                                i.getType() == Material.CHAINMAIL_BOOTS ||
                                                (!Utils.isBefore1_16() && i.getType() == Material.NETHERITE_BOOTS)) &&
                                                !i.getEnchantments().containsKey(Enchantment.BINDING_CURSE) &&
                                                e.getPlayer().getInventory().getBoots() == null)
                                            e.getPlayer().getInventory().setBoots(i);

                                        else if ((i.getType() == Material.IRON_CHESTPLATE ||
                                                i.getType() == Material.GOLDEN_CHESTPLATE ||
                                                i.getType() == Material.LEATHER_CHESTPLATE ||
                                                i.getType() == Material.DIAMOND_CHESTPLATE ||
                                                i.getType() == Material.CHAINMAIL_CHESTPLATE ||
                                                (!Utils.isBefore1_16() && i.getType() == Material.NETHERITE_BOOTS) ||
                                                i.getType() == Material.ELYTRA) &&
                                                !i.getEnchantments().containsKey(Enchantment.BINDING_CURSE) &&
                                                e.getPlayer().getInventory().getChestplate() == null)
                                            e.getPlayer().getInventory().setChestplate(i);

                                        else if ((i.getType() == Material.IRON_LEGGINGS ||
                                                i.getType() == Material.GOLDEN_LEGGINGS ||
                                                i.getType() == Material.LEATHER_LEGGINGS ||
                                                i.getType() == Material.DIAMOND_LEGGINGS ||
                                                i.getType() == Material.CHAINMAIL_LEGGINGS ||
                                                (!Utils.isBefore1_16() && i.getType() == Material.NETHERITE_LEGGINGS)) &&
                                                !i.getEnchantments().containsKey(Enchantment.BINDING_CURSE) &&
                                                e.getPlayer().getInventory().getLeggings() == null)
                                            e.getPlayer().getInventory().setLeggings(i);

                                        else if (e.getPlayer().getInventory().firstEmpty() != -1)
                                            e.getPlayer().getInventory().addItem(i);
                                        else
                                            e.getPlayer().getWorld().dropItemNaturally(block.getLocation(), i);
                                    }
                                }
                            } else {
                                // pushed item on the ground
                                for (ItemStack i : cd.getInventory()) {
                                    if (i != null) {
                                        e.getPlayer().getWorld().dropItemNaturally(block.getLocation(), i);
                                    }
                                }

                            }

                            block.setType(Material.AIR);
                            chestData.remove(cd);
                            fileManager.saveModification();
                            block.getWorld().playEffect(block.getLocation(), Effect.MOBSPAWNER_FLAMES, 10);
                            e.getPlayer().playSound(block.getLocation(), Sound.ENTITY_EXPERIENCE_ORB_PICKUP, 10, 1);
                            cd.removeArmorStand();
                            break;
                        } else {
                            e.setCancelled(true);
                            e.getPlayer().sendMessage(local.get("loc_prefix") + local.get("loc_not_owner"));
                        }
                    }
                }
            }
        }
    }

    @EventHandler
    public void onBlockBreakEvent(BlockBreakEvent e) {
        if (isGraveBlock(e.getBlock().getType())) {
            if (isIndestructible) {

                for (ChestData cd : chestData) {
                    if (cd.getChestLocation() == e.getBlock().getLocation()) {
                        e.setCancelled(true);
                        e.getPlayer().sendMessage(local.get("loc_prefix") + local.get("loc_not_owner"));
                        break;
                    }
                }
            }
        }
    }

    /**
     * Disable water destruction of Deadchest
     **/
    @EventHandler
    public void onBlockFromToEvent(BlockFromToEvent e) {
        if (isGraveBlock(e.getToBlock().getType())) {
            if (isIndestructible) {
                for (ChestData cd : chestData) {
                    if (cd.getChestLocation().equals(e.getToBlock().getLocation())) {
                        e.setCancelled(true);
                        break;
                    }
                }
            }
        }
    }


    @EventHandler
    public void onEntityExplodeEvent(EntityExplodeEvent e) {
        chestExplosionHandler(e);
    }

    @EventHandler
    public void onBlockExplodeEvent(BlockExplodeEvent e) {
        chestExplosionHandler(e);
    }

    public void chestExplosionHandler(Event e) {
        List<Block> blocklist = new ArrayList<>();
        if (e instanceof EntityExplodeEvent) {
            blocklist = ((EntityExplodeEvent) e).blockList();
        } else if (e instanceof BlockExplodeEvent) {
            blocklist = ((BlockExplodeEvent) e).blockList();
        }

        if (blocklist.size() > 0) {
            for (int i = 0; i < blocklist.size(); ++i) {
                Block block = blocklist.get(i);
                for (ChestData cd : chestData) {
                    if (isGraveBlock(block.getType()) && cd.getChestLocation().equals(block.getLocation())) {
                        if (isIndestructible) {
                            blocklist.remove(block);
                            generateLog("Deadchest of [" + cd.getPlayerName() + "] was protected from explosion in " + Objects.requireNonNull(cd.getChestLocation().getWorld()).getName());
                        } else {
                            cd.removeArmorStand();
                            chestData.remove(cd);
                            generateLog("Deadchest of [" + cd.getPlayerName() + "] was blown up in " + Objects.requireNonNull(cd.getChestLocation().getWorld()).getName());
                        }
                        break;
                    }
                }
            }
        }
    }

    @EventHandler
    public void onPlayerArmorStandManipulateEvent(PlayerArmorStandManipulateEvent e) {
        if (!e.getRightClicked().isVisible() && e.getRightClicked().getMetadata("deadchest").size() != 0) {
            e.setCancelled(true);
        }
    }

    @EventHandler
    public void onBlockPlaceEvent(BlockPlaceEvent e) {
        // Disable double chest for grave chest
        if (e.getBlock().getType() == Material.CHEST) {
            for (BlockFace face : BlockFace.values()) {
                Block block = e.getBlock().getRelative(face);
                if (block.getType() == Material.CHEST) {
                    for (ChestData cd : chestData) {
                        if (cd.getChestLocation().equals(block.getLocation())) {
                            e.setCancelled(true);
                            e.getPlayer().sendMessage(local.get("loc_prefix") + local.get("loc_doubleDC"));
                            return;
                        }
                    }
                }
            }
        }
    }
}

File: deadchest-core/src/main/kotlin/me/crylonz/deadchest/DeadChestListener.kt
package me.crylonz.deadchest

import me.crylonz.deadchest.DeadChest.Companion.chestData
import me.crylonz.deadchest.DeadChest.Companion.local
import me.crylonz.deadchest.Utils.generateLog
import me.crylonz.deadchest.Utils.isBoots
import me.crylonz.deadchest.Utils.isChestplate
import me.crylonz.deadchest.Utils.isHelmet
import me.crylonz.deadchest.Utils.isLeggings
import me.crylonz.deadchest.utils.ConfigKey
import me.crylonz.deadchest.utils.DeadChestConfig
import org.bukkit.*
import org.bukkit.block.Block
import org.bukkit.block.BlockFace
import org.bukkit.enchantments.Enchantment
import org.bukkit.entity.EntityType
import org.bukkit.entity.ExperienceOrb
import org.bukkit.event.Event
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.block.*
import org.bukkit.event.entity.EntityExplodeEvent
import org.bukkit.event.entity.PlayerDeathEvent
import org.bukkit.event.player.PlayerArmorStandManipulateEvent
import org.bukkit.event.player.PlayerInteractEvent
import org.bukkit.inventory.ItemStack
import java.util.*

class DeadChestListener(private val plugin: DeadChest) : Listener {

    val config: DeadChestConfig
        get() = DeadChest.dcConfig

    @EventHandler(priority = EventPriority.LOWEST)
    fun onPlayerDeathEvent(event: PlayerDeathEvent) {
        if (event.keepInventory) {
            return
        }
        val player = event.entity.player

        if (player == null
            || DeadChest.dcConfig.getArray(ConfigKey.EXCLUDED_WORLDS).contains(player.world.name)
            || !config.getBoolean(ConfigKey.GENERATE_DEADCHEST_IN_CREATIVE)
            && player.gameMode == GameMode.CREATIVE
        ) {
            return
        }
        if (Utils.worldGuardCheck(player) && (player.hasPermission(Permission.GENERATE.label)
                    || !config.getBoolean(ConfigKey.REQUIRE_PERMISSION_TO_GENERATE))
        ) {
            if ((DeadChestManager.playerDeadChestAmount(player) < config.getInt(ConfigKey.MAX_DEAD_CHEST_PER_PLAYER) ||
                        config.getInt(ConfigKey.MAX_DEAD_CHEST_PER_PLAYER) == 0) &&
                player.getMetadata("NPC").isEmpty()
            ) {
                val world = player.world
                var loc = player.location
                if (!config.getBoolean(ConfigKey.GENERATE_ON_LAVA) && loc.block.type == Material.LAVA) {
                    generateLog("Player dies in lava : No deadchest generated")
                    return
                }
                if (!config.getBoolean(ConfigKey.GENERATE_ON_WATER) && loc.block.type == Material.WATER) {
                    generateLog("Player dies in water : No deadchest generated")
                    return
                }
                if (!config.getBoolean(ConfigKey.GENERATE_ON_RAILS) && loc.block.type == Material.RAIL || loc.block.type == Material.ACTIVATOR_RAIL || loc.block.type == Material.DETECTOR_RAIL || loc.block.type == Material.POWERED_RAIL) {
                    generateLog("Player dies on rails : No deadchest generated")
                    return
                }
                if (!config.getBoolean(ConfigKey.GENERATE_IN_MINECART) && player.vehicle != null) {
                    if (player.vehicle!!.type == EntityType.MINECART) {
                        generateLog("Player dies in a minecart : No deadchest generated")
                        return
                    }
                }

                // Handle case bottom of the world
                val minHeight = Utils.computeMinHeight()
                if (loc.y < minHeight) {
                    loc.y = (world.getHighestBlockYAt(loc.x.toInt(), loc.z.toInt()) + 1).toDouble()

                    if (loc.y < minHeight) {
                        loc.y = minHeight.toDouble()
                    }
                } else if (loc.blockY >= world.maxHeight) {

                    var y = world.maxHeight - 1
                    loc.y = y.toDouble()

                    while (world.getBlockAt(loc).type != Material.AIR && y > 0) {
                        y--
                        loc.y = y.toDouble()
                    }

                    if (y < 1) {
                        player.sendMessage(local.get("loc_prefix") + local.get("loc_noDCG"))
                        return
                    }

                } else {
                    if (world.getBlockAt(loc).type == Material.DARK_OAK_DOOR ||
                        world.getBlockAt(loc).type == Material.ACACIA_DOOR ||
                        world.getBlockAt(loc).type == Material.BIRCH_DOOR ||
                        !Utils.isBefore1_16 && world.getBlockAt(loc).type == Material.CRIMSON_DOOR ||
                        world.getBlockAt(loc).type == Material.IRON_DOOR ||
                        world.getBlockAt(loc).type == Material.JUNGLE_DOOR ||
                        world.getBlockAt(loc).type == Material.OAK_DOOR ||
                        world.getBlockAt(loc).type == Material.SPRUCE_DOOR ||
                        !Utils.isBefore1_16 && world.getBlockAt(loc).type == Material.WARPED_DOOR ||
                        world.getBlockAt(loc).type == Material.VINE ||
                        world.getBlockAt(loc).type == Material.LADDER
                    ) {
                        val tmpLoc = Utils.getFreeBlockAroundThisPlace(world, loc)
                        if (tmpLoc != null) {
                            loc = tmpLoc
                        }
                    }
                    if (world.getBlockAt(loc).type != Material.AIR &&
                        world.getBlockAt(loc).type != Material.CAVE_AIR &&
                        world.getBlockAt(loc).type != Material.VOID_AIR &&
                        world.getBlockAt(loc).type != Material.WATER &&
                        world.getBlockAt(loc).type != Material.LAVA
                    ) {
                        while (world.getBlockAt(loc).type != Material.AIR && loc.y < world.maxHeight) {
                            loc.y = loc.y + 1
                        }
                    }
                }
                val groundLocation = loc.clone()
                groundLocation.y = groundLocation.y - 1
                if (Utils.isBefore1_17 && world.getBlockAt(groundLocation).type == Material.valueOf("GRASS_PATH") || !Utils.isBefore1_17 && world.getBlockAt(
                        groundLocation
                    ).type == Material.DIRT_PATH || world.getBlockAt(groundLocation).type == Material.FARMLAND
                ) {
                    loc.y = loc.y + 1
                }
                val b = world.getBlockAt(loc)
                if (!Utils.isInventoryEmpty(player.inventory)) {
                    Utils.computeChestType(b, player)
                    val firstLine =
                        local.replacePlayer(local.get("holo_owner"), event.entity.displayName)
                    val holoName = DeadChestManager.generateHologram(b.location, firstLine, 0.5f, -0.95f, 0.5f, false)
                    val secondLine = local.get("holo_loading")
                    val holoTime = DeadChestManager.generateHologram(b.location, secondLine, 0.5f, -1.2f, 0.5f, true)

                    // Remove items with curse of vanishing
                    for (itemStack: ItemStack? in player.inventory.contents) {
                        if (itemStack != null && itemStack.enchantments.containsKey(Enchantment.VANISHING_CURSE)) {
                            player.inventory.remove(itemStack)
                        }
                    }
                    if (player.inventory.helmet != null
                        && player.inventory.helmet!!.enchantments.containsKey(Enchantment.VANISHING_CURSE)
                    ) {
                        player.inventory.helmet = null
                    }
                    if (player.inventory.chestplate != null
                        && player.inventory.chestplate!!.enchantments.containsKey(Enchantment.VANISHING_CURSE)
                    ) {
                        player.inventory.chestplate = null
                    }
                    if (player.inventory.leggings != null
                        && player.inventory.leggings!!.enchantments.containsKey(Enchantment.VANISHING_CURSE)
                    ) {
                        player.inventory.leggings = null
                    }
                    if (player.inventory.boots != null
                        && player.inventory.boots!!.enchantments.containsKey(Enchantment.VANISHING_CURSE)
                    ) {
                        player.inventory.boots = null
                    }
                    if (player.inventory.itemInOffHand.enchantments.containsKey(Enchantment.VANISHING_CURSE)) {
                        player.inventory.setItemInOffHand(null)
                    }
                    for (item: String? in DeadChest.dcConfig.getArray(ConfigKey.EXCLUDED_ITEMS)) {
                        if (item != null && Material.getMaterial(item.uppercase(Locale.getDefault())) != null) {
                            player.inventory.remove(Material.getMaterial(item.uppercase(Locale.getDefault()))!!)
                        }
                    }
                    if (DeadChest.dcConfig.getBoolean(ConfigKey.STORE_XP)) {
                        event.droppedExp = 0
                    }
                    chestData.add(
                        ChestData(
                            player.inventory,
                            b.location,
                            player,
                            player.hasPermission(Permission.INFINITY_CHEST.label),
                            holoTime,
                            holoName,
                            Utils.computeXpToStore(player)
                        )
                    )
                    val backupInv = player.inventory.contents
                    event.drops.clear()
                    player.inventory.clear()
                    if (config.getBoolean(ConfigKey.DISPLAY_POSITION_ON_DEATH)) {
                        player.sendMessage(
                            local.get("loc_prefix") + local.get("loc_chestPos") + " X: " +
                                    ChatColor.WHITE + b.x + ChatColor.GOLD + " Y: " +
                                    ChatColor.WHITE + b.y + ChatColor.GOLD + " Z: " +
                                    ChatColor.WHITE + b.z
                        )
                    }
                    DeadChest.fileManager.saveModification()
                    generateLog("New deadchest for [" + player.name + "] in " + b.world.name + " at X:" + b.x + " Y:" + b.y + " Z:" + b.z)
                    generateLog("Chest content : " + Arrays.asList(*backupInv))
                    if (config.getBoolean(ConfigKey.LOG_DEADCHEST_ON_CONSOLE)) plugin.logger.info("New deadchest for [" + player.name + "] at X:" + b.x + " Y:" + b.y + " Z:" + b.z)
                } else {
                    generateLog("Player [" + player.name + "] died without inventory : No Deadchest generated")
                }
            }
        }
    }

    @EventHandler
    fun onClick(event: PlayerInteractEvent) {
        if ((event.action == Action.RIGHT_CLICK_BLOCK)) {
            chestData
                .filter {
                    it.chestLocation.world === event.player.world &&
                            it.chestLocation.distance(event.clickedBlock!!.location) <= 1
                }.forEach { _ ->
                    event.isCancelled = true
                    return@forEach
                }
        } else if (event.action == Action.LEFT_CLICK_BLOCK &&
            event.clickedBlock != null &&
            Utils.isGraveBlock(event.clickedBlock!!.type)
        ) {
            val block = event.clickedBlock!!
            val player = event.player
            val playerUUID = player.uniqueId.toString()
            val playerHasPermission = player.hasPermission(Permission.CHESTPASS.label)
            val playerWorld = player.world

            chestData
                .filter { it.chestLocation == block.location }
                .firstOrNull {

                    if (!config.getBoolean(ConfigKey.ONLY_OWNER_CAN_OPEN_CHEST) ||
                        playerUUID == it.playerUUID ||
                        playerHasPermission
                    ) {
                        if (!player.hasPermission(Permission.GET.label) &&
                            config.getBoolean(ConfigKey.REQUIRE_PERMISSION_TO_GET_CHEST)
                        ) {
                            generateLog("Player ${player.name} need to have deadchest.get permission to generate")
                            player.sendMessage(local.get("loc_prefix") + local.get("loc_noPermsToGet"))
                            event.isCancelled = true
                            return@onClick
                        }

                        val deadChestPickUpEvent = DeadChestPickUpEvent()
                        Bukkit.getServer().pluginManager.callEvent(DeadChestPickUpEvent())
                        if (!deadChestPickUpEvent.isCancelled) {
                            generateLog("Deadchest of [${it.playerName}] was taken by [${player.name}] in ${playerWorld.name}")

                            // put all item on the inventory
                            if (config.getInt(ConfigKey.DROP_MODE) == 1) {
                                val playerInventory = player.inventory
                                player.giveExp(it.xpStored)

                                it.inventory
                                    .filterNotNull()
                                    .forEach { item ->
                                        when {
                                            isHelmet(item) && playerInventory.helmet == null ->
                                                playerInventory.helmet = item

                                            isBoots(item) && playerInventory.boots == null ->
                                                playerInventory.boots = item

                                            isChestplate(item) && playerInventory.chestplate == null ->
                                                playerInventory.chestplate = item

                                            isLeggings(item) && playerInventory.leggings == null ->
                                                playerInventory.leggings = item

                                            playerInventory.firstEmpty() != -1 -> playerInventory.addItem(item)
                                            else -> playerWorld.dropItemNaturally(block.location, item)
                                        }
                                    }
                            } else {
                                // pushed item on the ground
                                it.inventory
                                    .filterNotNull()
                                    .forEach { item ->
                                        playerWorld.dropItemNaturally(block.location, item)

                                        if (it.xpStored != 0) {
                                            playerWorld.spawn(block.location, ExperienceOrb::class.java).run {
                                                experience = it.xpStored
                                            }
                                        }
                                    }
                            }

                            block.type = Material.AIR
                            chestData.remove(it)
                            DeadChest.fileManager.saveModification()
                            block.world.playEffect(block.location, Effect.MOBSPAWNER_FLAMES, 10)
                            player.playSound(block.location, Sound.ENTITY_EXPERIENCE_ORB_PICKUP, 10f, 1f)
                            it.removeArmorStand()
                            return@onClick
                        } else {
                            event.isCancelled = true
                        }
                    } else {
                        event.isCancelled = true
                        player.sendMessage(local.get("loc_prefix") + local.get("loc_not_owner"))
                    }
                    return
                }
        }
    }

    @EventHandler
    fun onBlockBreakEvent(event: BlockBreakEvent) {
        if (Utils.isGraveBlock(event.block.type)) {
            if (config.getBoolean(ConfigKey.INDESTRUCTIBLE_CHEST)) {
                for (cd: ChestData in chestData) {
                    if (cd.chestLocation === event.block.location) {
                        event.isCancelled = true
                        event.player.sendMessage(local.get("loc_prefix") + local.get("loc_not_owner"))
                        break
                    }
                }
            }
        }
    }

    /**
     * Disable water destruction of Deadchest (for head)
     */
    @EventHandler
    fun onBlockFromToEvent(event: BlockFromToEvent) {
        if (Utils.isGraveBlock(event.toBlock.type)) {
            for (cd: ChestData in chestData) {
                if ((cd.chestLocation == event.toBlock.location)) {
                    event.isCancelled = true
                    break
                }
            }
        }
    }

    @EventHandler
    fun onEntityExplodeEvent(event: EntityExplodeEvent?) {
        chestExplosionHandler(event)
    }

    @EventHandler
    fun onBlockExplodeEvent(event: BlockExplodeEvent?) {
        chestExplosionHandler(event)
    }

    private fun chestExplosionHandler(event: Event?) {

        var blocklist: MutableList<Block> = ArrayList()
        if (event is EntityExplodeEvent) {
            blocklist = event.blockList()
        } else if (event is BlockExplodeEvent) {
            blocklist = event.blockList()
        }
        if (blocklist.size > 0) {

            blocklist.forEach { block ->
                chestData
                    .filter { Utils.isGraveBlock(block.type) && it.chestLocation == block.location }
                    .forEach {
                        if (config.getBoolean(ConfigKey.INDESTRUCTIBLE_CHEST)) {
                            blocklist.remove(block)
                            generateLog("Deadchest of [" + it.playerName + "] was protected from explosion in " + it.chestLocation.world!!.name)
                        } else {
                            it.removeArmorStand()
                            chestData.remove(it)
                            generateLog(
                                "Deadchest of [" + it.playerName + "] was blown up in " + it.chestLocation.world!!.name
                            )
                        }
                    }
            }
        }

        @EventHandler
        fun onPlayerArmorStandManipulateEvent(event: PlayerArmorStandManipulateEvent) {
            if (!event.rightClicked.isVisible && event.rightClicked.getMetadata("deadchest").size != 0) {
                event.isCancelled = true
            }
        }

        @EventHandler
        fun onBlockPlaceEvent(event: BlockPlaceEvent) {
            // Disable double chest for grave chest
            if (event.block.type == Material.CHEST) {

                BlockFace.values()
                    .map { event.block.getRelative(it) }
                    .filter { it.type == Material.CHEST }
                    .forEach { block ->
                        chestData
                            .filter { it.chestLocation == block.location }
                            .forEach { _ ->
                                event.isCancelled = true
                                event.player.sendMessage(local.get("loc_prefix") + local.get("loc_doubleDC"))
                                return@onBlockPlaceEvent
                            }
                    }
            }
        }

        @EventHandler(priority = EventPriority.HIGH)
        fun onBlockPistonExtendEvent(event: BlockPistonExtendEvent) {

            event.blocks
                .filterNotNull()
                .filter { it.type == Material.PLAYER_HEAD || it.type == Material.PLAYER_WALL_HEAD }
                .forEach { block ->
                    chestData
                        .filter { it.chestLocation == block.location }
                        .forEach { _ ->
                            event.isCancelled = true
                            return@onBlockPistonExtendEvent
                        }
                }
        }
    }
}
